client/relay-chain-minimal-node/src/network.rs

```diff
@@ -0,0 +1,420 @@
+	/// task and notify later when more actions are ready to be polled. To continue the comparison,
+	/// it is as if this method always returned `Poll::Pending`.
+	fn poll_actions(
+		&mut self,
+		_cx: &mut futures::task::Context,
+		_link: &mut dyn sc_consensus::import_queue::Link<Block>,
+	) {
+	}
+}
+
+/// Build the network service, the network status sinks and an RPC sender.
+pub(crate) fn build_collator_network(
+	params: BuildCollatorNetworkParams,
+) -> Result<(Arc<NetworkService<Block, Hash>>, NetworkStarter), Error> {
+	let BuildCollatorNetworkParams { config, client, spawn_handle, genesis_hash } = params;
+
+	let transaction_pool_adapter = Arc::new(sc_network::config::EmptyTransactionPool {});
+
+	let protocol_id = config.protocol_id();
+
+	let block_request_protocol_config =
+		block_request_handler::generate_protocol_config(&protocol_id, genesis_hash, None);
+
+	let state_request_protocol_config =
+		state_request_handler::generate_protocol_config(&protocol_id, genesis_hash, None);
+
+	let light_client_request_protocol_config =
+		light_client_requests::generate_protocol_config(&protocol_id, genesis_hash, None);
+
+	let network_params = sc_network::config::Params {
+		role: config.role.clone(),
+		executor: {
+			let spawn_handle = Clone::clone(&spawn_handle);
+			Some(Box::new(move |fut| {
+				spawn_handle.spawn("libp2p-node", Some("networking"), fut);
+			}))
+		},
+		transactions_handler_executor: {
+			let spawn_handle = Clone::clone(&spawn_handle);
+			Box::new(move |fut| {
+				spawn_handle.spawn("network-transactions-handler", Some("networking"), fut);
+			})
+		},
+		fork_id: None,
+		chain_sync: Box::new(DummyChainSync),
+		network_config: config.network.clone(),
+		chain: client.clone(),
+		transaction_pool: transaction_pool_adapter as _,
+		import_queue: Box::new(DummyImportQueue),
+		protocol_id,
+		metrics_registry: config.prometheus_config.as_ref().map(|config| config.registry.clone()),
+		block_request_protocol_config,
+		state_request_protocol_config,
+		warp_sync_protocol_config: None,
+		light_client_request_protocol_config,
+		request_response_protocol_configs: Vec::new(),
+	};
+
+	let network_mut = sc_network::NetworkWorker::new(network_params)?;
+	let network = network_mut.service().clone();
+
+	let future = build_network_collator_future(network_mut);
+
+	let (network_start_tx, network_start_rx) = futures::channel::oneshot::channel();
+
+	// The network worker is responsible for gathering all network messages and processing
+	// them. This is quite a heavy task, and at the time of the writing of this comment it
+	// frequently happens that this future takes several seconds or in some situations
+	// even more than a minute until it has processed its entire queue. This is clearly an
+	// issue, and ideally we would like to fix the network future to take as little time as
+	// possible, but we also take the extra harm-prevention measure to execute the networking
+	// future using `spawn_blocking`.
```

Comment:
So the processing of the queue can take more than a minute, and it spends most of this time *actually* burning CPU time instead of waiting? If so that does sound quite abnormal to me.

Reply:
I copied this part of the code from [here](https://github.com/paritytech/substrate/blob/61b9a4d1a8a9bf39c1d89a8dd02f82785c10860c/client/service/src/builder.rs#L931), I did not investigate this specific issue. Could very well be that this is outdated.

Reply:
I see. Okay, then you can leave it as-is.

From what I can see @tomaka added this over two years ago, so it's probably worth revisit.

Reply:
I'm not following networking code changes, but given the lack of huge changes I think that this is still accurate.
I don't know what you have in mind when you say "burning" CPU, but yes we're talking about several seconds of CPU cycles to run `poll()` once.